Add StackTraceLimit and SpanFramesMinDurationInMilliseconds configs

gregkalapos opened the pull request:
Solves elastic/apm-agent-dotnet#307

Added `SpanFramesMinDurationInMilliseconds` and `StackTraceLimit`. Those work the same as in other agents.

Main benefit:
- By setting one of those setting to 0 users can disable stack trace collection, which causes perf. overhead. For numbers [see this comment](https://github.com/elastic/apm-agent-dotnet/issues/307#issuecomment-510134603).

The `StackTraceLimit` is implemented by manually trimming the whole stack trace - so there is no perf difference between capturing only 1 stack frame vs. capturing all frames (see [discussion here](https://github.com/elastic/apm-agent-dotnet/issues/307#issuecomment-510202556)), nevertheless with this we are still in sync with other agents in terms of configs and with this PR we give the users the option to control stack frame collection and they can turn it off by setting those configs to `0`.

gregkalapos commented:
Fighting with CI - it seems that in CI we don't get line number info and from that reason we have 1 failing test.

Can't repro locally - trying to figure this out - I don't see yet how this PR changes anything on that part 🤷‍♂️

SergeyKleyman commented:
> Fighting with CI - it seems that in CI we don't get line number info and from that reason we have 1 failing test.
> 
> Can't repro locally - trying to figure this out - I don't see yet how this PR changes anything on that part 🤷‍♂️

I see that the test fails only on Linux - is it possible that on Linux assemblies don't have debugging info with file names and line numbers? It's not clear why this started to happen with this PR but maybe CI machines when through some changes? What would be the simplest way to re-run tests on the current master on CI machines?

SergeyKleyman commented on test/Elastic.Apm.Tests/StackTraceTests.cs:
```diff
@@ -27,6 +28,9 @@ public class StackTraceTests
 		public async Task HttpClientStackTrace()
 		{
+			if(RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+				return; //in CI on Windows line numbers don't seem to be available.
```
Are you sure? Based on https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fapm-agent-dotnet-mbp/detail/PR-374/3/tests/

> Existing failures - 1
> Initializing / Parallel / Linux / Test / HttpClientStackTrace – Elastic.Apm.Tests.StackTraceTests

Tests fail only on Linux.

gregkalapos replied:
Yeah, they fail both on Linux and Windows. It was just not clear to me previously that they also fail on Linux - somehow Jenkins showed those differently. Anyway, in https://github.com/elastic/apm-agent-dotnet/pull/374/commits/69d196ebee5cffe6d1b0e1fd8ea8639383251671 I tried to disable it for windows - with that it's now clear that it fails everywhere.

gregkalapos replied:
> > Fighting with CI - it seems that in CI we don't get line number info and from that reason we have 1 failing test.
> > Can't repro locally - trying to figure this out - I don't see yet how this PR changes anything on that part 🤷‍♂️
> 
> I see that the test fails only on Linux - is it possible that on Linux assemblies don't have debugging info with file names and line numbers? It's not clear why this started to happen with this PR but maybe CI machines when through some changes? What would be the simplest way to re-run tests on the current master on CI machines?

I merged something today, and with that all tests were green: https://github.com/elastic/apm-agent-dotnet/commit/0b66a3212e9df7b01c29752469cdb4b1b8efb45d (only thing made it red is that the test coverage is below target, but that's a known bug and the next build will make it good again. - I think there is no change between this branch and master in terms of jenkins and CI related things.

gregkalapos commented:
> > Fighting with CI - it seems that in CI we don't get line number info and from that reason we have 1 failing test.
> > Can't repro locally - trying to figure this out - I don't see yet how this PR changes anything on that part 🤷‍♂️
> 
> I see that the test fails only on Linux - is it possible that on Linux assemblies don't have debugging info with file names and line numbers? It's not clear why this started to happen with this PR but maybe CI machines when through some changes? What would be the simplest way to re-run tests on the current master on CI machines?

I merged something today, and with that all tests were green: https://github.com/elastic/apm-agent-dotnet/commit/0b66a3212e9df7b01c29752469cdb4b1b8efb45d (only thing made it red is that the test coverage is below target, but that's a known bug and the next build will make it good again. - I think there is no change between this branch and master in terms of jenkins and CI related things.

gregkalapos commented on src/Elastic.Apm/Config/AbstractConfigurationReader.cs:
```diff
@@ -54,14 +54,17 @@ protected string ParseSecretToken(ConfigurationKeyValue kv)
-			if (bool.TryParse(kv.Value, out var value)) return value;
 
-			return true;
+			return !bool.TryParse(kv.Value, out var value) || value;
```
This was auto cleanup - if we revert back then the next auto cleanup will make it like this again.

SergeyKleyman replied:
You might be using too aggressive option of Rider's code cleanup. I use `SilentCodeCleanUp`:
![image](https://user-images.githubusercontent.com/7782093/61325896-5d575300-a81e-11e9-9ce9-90991135eaeb.png)

gregkalapos replied:
Yeah, I usually do full cleanup.

Plus to me it also suggests this change - I thought it's maybe based on our [`DotSettings` file](https://github.com/elastic/apm-agent-dotnet/blob/master/ElasticApmAgent.sln.DotSettings)  - but I'm not sure

<img width="634" alt="image" src="https://user-images.githubusercontent.com/1091853/61326456-174ac100-a817-11e9-9373-9842cd4d8759.png">

To me the green dots and the yellow thing was a little bit disturbing. Do you still want me to revert?

SergeyKleyman replied:
Not a major thing for me - up to you. if you decide to revert, as I mentioned above, I would find it more useful to replace `true` in the code before the change with a named constant `DefaultValues.CaptureHeaders`.

gregkalapos replied:
It would be nice to know why it's suggested - do you also see this suggestion? I think if it's defined in our `DotSettings` we should do it this way, if not I'd also not do this.

SergeyKleyman replied:
Yes, I also see the suggestion. I copied the solution aside, deleted `.editorconfig` and I still see the same suggestion.

gregkalapos replied:
Visual Studio also gives a suggestion, but it suggest an (in my opinion) more readable version.

<img width="1139" alt="image" src="https://user-images.githubusercontent.com/1091853/61327728-de601b80-a819-11e9-96b4-094c0639129e.png">

But once I take it, Rider still suggests the other one 🤦‍♂

SergeyKleyman replied:
I tried to introduce a name constant `ConfigConsts.DefaultValues.CaptureHeaders` instead of `true` but Rider still suggest the same transformation because it sees the actual value of the constant:
```
return !bool.TryParse(kv.Value, out var value) || value;
```

 If I change `CaptureHeaders` to `static readonly` instead of `const` Rider still suggest to change `if` to `return` but at least it keeps `CaptureHeaders` in the picture
```
return bool.TryParse(kv.Value, out var value) ? value : ConfigConsts.DefaultValues.CaptureHeaders;
```

Which to me shows that using these suggestions automatically is too error prone and the benefit doesn't outweigh the risk. I frankly think that the first transformation is a bug because it takes the constant out of the picture so changing the constant after that will introduce a bug.

This particular change is not that important so the decision is yours but if I were you I would reconsider applying these changes automatically.

gregkalapos commented on test/Elastic.Apm.Tests/StackTraceTests.cs:
```diff
@@ -175,18 +176,260 @@ public void InheritedChain()
+				payloadSender.FirstSpan.StackTrace.Should().NotBeEmpty();
+				payloadSender.FirstSpan.StackTrace.Should().HaveCountGreaterThan(10);
+
+				//We don't filter out frames from the agent, therefore the current implementation will contain 2 agent frames on top
```
I asked, and no agent was mentioned where agent frames are filtered out. I think it's something worth discussing, but I'd keep it this way in this PR.

gregkalapos commented on src/Elastic.Apm/Config/AbstractConfigurationReader.cs:
```diff
@@ -54,14 +54,17 @@ protected string ParseSecretToken(ConfigurationKeyValue kv)
-			if (TryParseLogLevel(kv?.Value, out var level)) return level.Value;
+			if (TryParseLogLevel(kv?.Value, out var level))
+			{
+				if (level != null)
```
That's all true, but it still causes a warning:

<img width="1149" alt="image" src="https://user-images.githubusercontent.com/1091853/61326779-d2735a00-a817-11e9-9250-5bd5070d05e2.png">

I think this is a really minor change and fine this way - I just wanted to get rid of the warning.

SergeyKleyman replied:
The warning is indeed relevant because from `out var level` `level` is deduced to be `LogLevel?` and indeed `Nullable<T>.Value` can throw `InvalidOperationException`. I think the better solution to this issue would be to make `level` parameter of `TryParseLogLevel` non-nullable as I suggested above.

Not a major point - you decide.

gregkalapos commented:
The `StackFrame` class has a very promising feature - see [here](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.stackframe.-ctor?view=netframework-4.8#System_Diagnostics_StackFrame__ctor_System_Int32_) 

With that constructor it's possible to create single stack frames instead of getting the whole stack trace with `new StackTrace(true).GetFrames()`. 

Which sounds very promising, since e.g. if we only need 10 frames we could just create the 10 `StackFrame` instances. 

So, something like this in `Span.End()`:

```
if (_configurationReader.StackTraceLimit > 0)
{
	var stackFrames = new StackFrame[_configurationReader.StackTraceLimit];
	for (var i = 0; i < _configurationReader.StackTraceLimit; i++)
	{
		stackFrames[i] = new StackFrame(i);
	}
	StackTrace = StacktraceHelper.GenerateApmStackTrace(stackFrames, _logger,
		_configurationReader, $"Span `{Name}'");
}
else
{
	StackTrace = StacktraceHelper.GenerateApmStackTrace(new StackTrace(true).GetFrames(), _logger,
		_configurationReader, $"Span `{Name}'");
}
```

I implemented this, but it does not seem to help.

Test case: 100 Transactions each with 10Spans, `spanFramesMinDurationInMilliseconds: "-1ms", stackTraceLimit: "10"`.

Original code (no manual `StackFrame` creation, only `new StackTrace(true).GetFrames()`:

```
|                      Method |     Mean |    Error |   StdDev | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|---------------------------- |---------:|---------:|---------:|------------:|------------:|------------:|--------------------:|
| Simple100Transaction10Spans | 65.61 ms | 1.300 ms | 1.216 ms |   5375.0000 |   1250.0000 |           - |            32.95 MB |
```

Adding this "optimization":

```
|                      Method |     Mean |    Error |   StdDev | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|---------------------------- |---------:|---------:|---------:|------------:|------------:|------------:|--------------------:|
| Simple100Transaction10Spans | 132.0 ms | 1.231 ms | 1.092 ms |   6500.0000 |   1000.0000 |           - |             40.1 MB |
``` 

Clearly no improvement, won't commit 😔

gregkalapos commented on src/Elastic.Apm/Config/AbstractConfigurationReader.cs:
```diff
@@ -202,7 +256,7 @@ protected virtual string DiscoverServiceName()
 			foreach (var frame in stackFrames)
 			{
-				var currentAssemblyName = frame?.GetMethod()?.DeclaringType?.Assembly?.GetName();
+				var currentAssemblyName = frame?.GetMethod()?.DeclaringType?.Assembly.GetName();
```
It looks like that... this was also a warning previously: 

<img width="1035" alt="image" src="https://user-images.githubusercontent.com/1091853/61332895-b8407880-a825-11e9-9aa2-089a466b92a8.png">

SergeyKleyman replied:
As I mentioned in another comment I would be careful with these Rider's suggestions. Is there an explanation as to why `Type.Assembly` cannot return `null`? If there's no convincing explanation then is it be better to override this Rider's warning and keep the check thus avoiding `NullReferenceException` or trust Rider? The decision is yours.

gregkalapos commented on src/Elastic.Apm/Config/AbstractConfigurationReader.cs:
```diff
@@ -186,7 +226,21 @@ private bool TryParseTimeInterval(string valueAsString, out double valueInMillis
+					switch (defaultSuffix)
+					{
+						case TimeSuffix.M:
+							valueInMilliseconds = TimeSpan.FromMinutes(valueInSecondsNoUnits).TotalMilliseconds;
```
👍  - done

gregkalapos commented on src/Elastic.Apm/Config/AbstractConfigurationReader.cs:
```diff
@@ -186,7 +226,21 @@ private bool TryParseTimeInterval(string valueAsString, out double valueInMillis
+						case TimeSuffix.S:
+							valueInMilliseconds = TimeSpan.FromSeconds(valueInSecondsNoUnits).TotalMilliseconds;
+							break;
+						default: throw new ArgumentOutOfRangeException(nameof(defaultSuffix), defaultSuffix, null);
```
Yes, that was my intention - fixed, thx!

gregkalapos commented on src/Elastic.Apm/Config/ConfigConsts.cs:
```diff
@@ -4,30 +4,35 @@ namespace Elastic.Apm.Config
-			public const string CaptureHeaders = "ELASTIC_APM_CAPTURE_HEADERS";
-			public const string TransactionSampleRate = "ELASTIC_APM_TRANSACTION_SAMPLE_RATE";
-			public const string MetricsInterval = "ELASTIC_APM_METRICS_INTERVAL";
+			public const double MinMetricsIntervalInMillisecond = 1000;
```
👍  done

gregkalapos commented on src/Elastic.Apm/Config/ConfigConsts.cs:
```diff
@@ -4,30 +4,35 @@ namespace Elastic.Apm.Config
+			public const string MetricsInterval = "30s";
+			public const double MetricsIntervalInMilliseconds = 30 * 1000;
+			public const string SpanFramesMinDuration = "5ms";
+			public const double SpanFramesMinDurationInMilliseconds = 5 * 1000;
```
Nice catch - fixed, thx!

gregkalapos commented on src/Elastic.Apm/Config/EnvironmentConfigurationReader.cs:
```diff
@@ -10,19 +10,23 @@ internal class EnvironmentConfigurationReader : AbstractConfigurationReader, ICo
 		public LogLevel LogLevel => ParseLogLevel(Read(ConfigConsts.EnvVarNames.LogLevel));
 
-		public IReadOnlyList<Uri> ServerUrls => ParseServerUrls(Read(ConfigConsts.EnvVarNames.ServerUrls));
+		public double MetricsIntervalInMillisecond => ParseMetricsInterval(Read(ConfigConsts.EnvVarNames.MetricsInterval));
```
fixed

gregkalapos commented on src/Elastic.Apm/Config/IAgentConfigReader.cs:
```diff
@@ -6,12 +6,29 @@ namespace Elastic.Apm.Config
+
+		/// <summary>
+		/// The number of stack frames the agent collects.
+		/// 0: no stacktrace is collected
```
I added some more comments there covering this.

gregkalapos commented on src/Elastic.Apm/Helpers/StacktraceHelper.cs:
```diff
@@ -20,9 +21,26 @@ internal static class StacktraceHelper
+			IConfigurationReader configurationReader, string dbgCapturingFor
+		)
 		{
+			var stackTraceLimit = configurationReader.StackTraceLimit;
```
Yeah, that's a good point. I changed this, we only read it only once. For now I think we are good with this, and we already have open issues for hot changeable configs, so this will probably evolve later.

gregkalapos commented on src/Elastic.Apm/Model/ExecutionSegmentCommon.cs:
```diff
@@ -156,10 +158,12 @@ internal static Task<T> CaptureSpan<T>(Span span, Func<ISpan, Task<T>> func)
-					$"{nameof(Transaction)}.{nameof(CaptureException)}")
 			};
 
+			if (configurationReader.StackTraceLimit != 0)
```
Yeah, it was on purpose, but it doesn't really matter - removed it.

gregkalapos commented on test/Elastic.Apm.Tests/ConfigTests.cs:
```diff
@@ -356,12 +362,42 @@ public void SetMetricsIntervalToNegativeMilliseconds()
 		/// </summary>
 		[Fact]
 		public void MetricsIntervalDefaultValuesInSync()
 			=> MetricsIntervalTestCommon(DefaultValues.MetricsInterval).Should().Be(DefaultValues.MetricsIntervalInMilliseconds);
```
Added

gregkalapos commented on src/Elastic.Apm/Model/Span.cs:
```diff
@@ -17,6 +18,7 @@ internal class Span : ISpan
 		private readonly Transaction _enclosingTransaction;
 		private readonly IApmLogger _logger;
 		private readonly IPayloadSender _payloadSender;
+		private readonly IConfigurationReader _configurationReader;
```
I'd skip on this one - we pass the _config to other methods, like `ExecutionSegmentCommon` and I think there are some other ones. I'd follow the same pattern and not share the instance through a reference to `Transaction`.

gregkalapos commented on test/Elastic.Apm.Tests/StackTraceTests.cs:
```diff
@@ -175,18 +176,260 @@ public void InheritedChain()
+			}
+
+			payloadSender.FirstSpan.Should().NotBeNull();
+			payloadSender.FirstSpan.StackTrace.Should().BeNullOrEmpty();
```
Because this tests the default config values and the assumption is that the empty span will be less than 5ms, therefore no stacktrace will be captured. (more in the next comment :) )

SergeyKleyman replied:
What I meant is that if stack trace is not captured then it will be null but how it can be empty? So shouldn't the line
```
payloadSender.FirstSpan.StackTrace.Should().BeNullOrEmpty();
```
be 
```
payloadSender.FirstSpan.StackTrace.Should().BeNull();
```
?

gregkalapos replied:
Och, I see. 

Yes, that's correct, but I thought it's more like an implementation detail - it doesn't really matter if it's an empty `List` or `null` - the point is that there are no stack frames. Seems more robust to me.

SergeyKleyman replied:
But stack trace cannot be empty by definition - stack trace for a span contains at least the application entry method (`Program.Main`) and stack trace for an error contains at least the method where exception is caught.

gregkalapos replied:
Yes, but the meaning of "no stack trace captured" can be represented as a list containing no item. - sometimes people want to avoid `null`. We don't currently do that, and I think that's good, but that's just an implementation detail.

But this is really a minor detail - if you want I can change it. I personally think it's more robust since I don't wanted to test on `null`, I wanted to test on not having stack frames, but I'm fine with changing it.

SergeyKleyman replied:
Not a major point - up to you. I understand that empty container is sometimes better than `null` as Null Object pattern application (like for example empty C# `event` without handlers) but I think in this case `null` is clearer.

In any case it might be worth making sure it's consistent with the rest of of the test cases.

gregkalapos commented on test/Elastic.Apm.Tests/StackTraceTests.cs:
```diff
@@ -175,18 +176,260 @@ public void InheritedChain()
+				agent.Tracer.CaptureTransaction("TestTransaction", "Test", t
+					=>
+				{
+					t.CaptureSpan("span", "span", () => { });
```
I don't think so... it's an empty lambda, I think it's safe to assume that it'll be executed in 5 ms. 

> Maybe it's worth changing the configuration value to something much higher than 5 ms?

The point of this test is to test the default configs, so I'd prefer to keep it - seems to be stable in CI.

gregkalapos commented on test/Elastic.Apm.Tests/StackTraceTests.cs:
```diff
@@ -175,18 +176,260 @@ public void InheritedChain()
+
+			payloadSender.FirstSpan.Should().NotBeNull();
+			payloadSender.FirstSpan.StackTrace.Should().NotBeNullOrEmpty();
+			payloadSender.FirstSpan.StackTrace.Should().HaveCount(ConfigConsts.DefaultValues.StackTraceLimit);
```
Right, this only worked, because XUnit with Rider already creates a stack trace that is deep enough. I removed this assert.

gregkalapos commented on test/Elastic.Apm.Tests/StackTraceTests.cs:
```diff
@@ -175,18 +176,260 @@ public void InheritedChain()
+		}
+
+		[Fact]
+		public void ErrorWithDefaultStackTrace2()
```
Right, fixed, thx!

gregkalapos commented on test/Elastic.Apm.Tests/StackTraceTests.cs:
```diff
@@ -175,18 +176,260 @@ public void InheritedChain()
+		/// </summary>
+		/// <exception cref="Exception"></exception>
+		[Fact]
+		public void ErrorWithDefaultStackTrace2WithSpanFramesMinDurationZero()
```
fixed

gregkalapos commented:
> > > Fighting with CI - it seems that in CI we don't get line number info and from that reason we have 1 failing test.
> > > Can't repro locally - trying to figure this out - I don't see yet how this PR changes anything on that part 🤷‍♂️
> > 
> > 
> > I see that the test fails only on Linux - is it possible that on Linux assemblies don't have debugging info with file names and line numbers? It's not clear why this started to happen with this PR but maybe CI machines when through some changes? What would be the simplest way to re-run tests on the current master on CI machines?
> 
> I merged something today, and with that all tests were green: [0b66a32](https://github.com/elastic/apm-agent-dotnet/commit/0b66a3212e9df7b01c29752469cdb4b1b8efb45d) (only thing made it red is that the test coverage is below target, but that's a known bug and the next build will make it good again.) - I think there is no change between this branch and master in terms of jenkins and CI related things.

Sooo.. I think the `.pdb`s are simply missing. Also that test is a little bit unfortunate, because it assumes that `.pdb`s are present, but that does not necessarily have to be the case. 

I'm thinking about dropping that test, since it causes more trouble than added value.

SergeyKleyman commented:
> Sooo.. I think the `.pdb`s are simply missing. Also that test is a little bit unfortunate, because it assumes that `.pdb`s are present, but that does not necessarily have to be the case.
> 
> I'm thinking about dropping that test, since it causes more trouble than added value.

The question is why `.pdb`s are missing only after changes in this PR?

I would not recommend removing tests before understanding why they fail - to me it contradicts the whole purpose of the tests. Also I would not recommend deferring investigation to a separate issue since it's very risky to merge this PR before understanding why tests fail. But it's your decision to make.

gregkalapos commented:
> > Sooo.. I think the `.pdb`s are simply missing. Also that test is a little bit unfortunate, because it assumes that `.pdb`s are present, but that does not necessarily have to be the case.
> > I'm thinking about dropping that test, since it causes more trouble than added value.
> 
> The question is why `.pdb`s are missing only after changes in this PR?
> 
> I would not recommend removing tests before understanding why they fail - to me it contradicts the whole purpose of the tests. Also I would not recommend deferring investigation to a separate issue since it's very risky to merge this PR before understanding why tests fail. But it's your decision to make.

Already talking to @v1v. We try to figure this out. 

Nevertheless the test skipped if there are no `.pdb` files - that's also the prod. behaviour - if there is no pdb we can't show line numbers, so the test is only valid if there are `.pdb`s.

SergeyKleyman commented on src/Elastic.Apm/Config/AbstractConfigurationReader.cs:
```diff
@@ -234,7 +237,9 @@ private bool TryParseTimeInterval(string valueAsString, out double valueInMillis
-						default: throw new ArgumentException( "Unexpected TimeSuffix value", nameof(defaultSuffix));
+						default:
+							valueInMilliseconds = -1;
+							return false;
```
Are you sure you want to just swallow incorrect input (which most likely means a bug in agents code)? For example [TryParse(String, NumberStyles, IFormatProvider, Int32) ](https://docs.microsoft.com/en-us/dotnet/api/system.int32.tryparse?view=netframework-4.8#System_Int32_TryParse_System_String_System_Globalization_NumberStyles_System_IFormatProvider_System_Int32__)
> ArgumentException
> style is not a NumberStyles value.

gregkalapos replied:
I reverted it, but if we throw an `ArgumentException` then we have to handle it - not being able to parse these values should not crash the agent even if it's an agent bug (I hope we agree at least on this).

If that's the case then we should handle it in the caller method, which is basically the same as when the method returns false - so from a single case, which is not being able to parse and returning false we made 2 cases: 1) returning false, 2) exception. Now we have to handle both, but the handling logic is the same. 

Here is the commit: https://github.com/elastic/apm-agent-dotnet/pull/374/commits/cddbba592d414cce423c4546667ee729fe0c70d9 - I think this code is very noise and repeats the same.

Nevertheless I think this is not that important - let's keep it this way.

SergeyKleyman replied:
It's not the same - units cannot be affected by user's input that means if the units are invalid it's a bug in agent's code (or maybe some mix of assemblies from different incompatible versions, etc.) Unlike invalid configuration option value which is just an error in logging terms, the bug with invalid units is a critical event. I think in case of a critical event there's no recovery and fail-fast approach should be used which for application it usually means record the current state to help with investigation and exit as soon as possible to prevent further possibly much more disastrous business data corruption. For the agent the second part is trickier - we can consider if killing the whole application might be an option users might want to have but at the very least agent should definitely stop doing ANYTHING.

gregkalapos replied:
> not being able to parse these values should not crash the agent even if it's an agent bug (I hope we agree at least on this).

Ok, we don't even agree on that one. 🤷‍♂️

I think it makes sense to not spend too much on it. I'll do whichever you want. Let me know if you want further changes.

SergeyKleyman replied:
I think you just misread my note - I'm not talking about user's input which of course can contain invalid units, I'm talking about parameters passed by agent's code which should be valid unless there is a bug.

gregkalapos replied:
Yes, I know that. I just think we overcomplicate things.

Adapting code to your suggestion.

SergeyKleyman replied:
Regarding the way to handle - I think logging critical event is okay for now. I added the link to this discussion to https://github.com/elastic/apm-agent-dotnet/issues/208 which I think should include minimal infrastructure to handle agent's bugs (fail-fast approach or maybe something else). We can discuss it when we get to #208 (not 1.0)

SergeyKleyman commented:
> Nevertheless the test skipped if there are no `.pdb` files - that's also the prod. behaviour - if there is no pdb we can't show line numbers, so the test is only valid if there are `.pdb`s.

I'm not sure I understand what you mean - if there are no line numbers at prod then why do we need all the agent production code and the tests dealing with the case when there are line numbers available?

gregkalapos commented:
> > Nevertheless the test skipped if there are no `.pdb` files - that's also the prod. behaviour - if there is no pdb we can't show line numbers, so the test is only valid if there are `.pdb`s.
> 
> I'm not sure I understand what you mean - if there are no line numbers at prod then why do we need all the agent production code and the tests dealing with the case when there are line numbers available?

Maybe I wrote too quickly. 

So, if there are no pdb files, even with `new StackTrace(true)` we won't get this line number information. 
So there are 2 cases:
- pdbs are available for the application binaries: in this case we deliver line numbers and I think we should test this.
- no pdbs available: in this case we won't be able to deliver line numbers.

My suggestion is that the tests also adapts to this and it will be skipped when there are no pdbs, instead of failing. The reasoning is that without pdbs making the tests fail would be false positive, since there is no way to make the feature work without the pdbs... that's not a real failure. 

Still I agree that we should investigate this and make sure we test this in CI - we are working on that.

SergeyKleyman commented:
> My suggestion is that the tests also adapts to this and it will be skipped when there are no pdbs, instead of failing. The reasoning is that without pdbs making the tests fail would be false positive, since there is no way to make the feature work without the pdbs... that's not a real failure.

No problem with having the test adapt but then we need to make sure that CI builds a separate variant with and without .pdb-s if we want to make that both work.

SergeyKleyman commented:
Also the way test adapts cannot self-referential (namely if there are no line numbers it means there are no .pdb-s) because then any bug affecting line number capture will be swallowed. The indication whether there are .pdb-s or there aren't should come from CI that knows which variant it built (for example via env var).

gregkalapos commented:
> Also the way test adapts cannot self-referential (namely if there are no line numbers it means there are no .pdb-s) because then any bug affecting line number capture will be swallowed. The indication whether there are .pdb-s or there aren't should come from CI that knows which variant it built (for example via env var).

Yeah, ok. I give up on this, won't do it...

gregkalapos commented:
> I don't recommend merging this PR before understanding why some tests started to fail at CI.

It was a single test (`HttpClientStackTrace`). Fixed now 🎉 , investigation with some details on #381.

SergeyKleyman commented:
Please see https://github.com/elastic/apm-agent-dotnet/issues/381#issuecomment-512691833

gregkalapos commented:
Planning to merge this (#381 with more details about the tests).

@v1v I removed the debugging stuff from the `jenkinsfile` - I left the `'find . -name *.pdb'` I think that can be generally useful, the other things were just to archive the pdbs, which gave me the info I needed. Added you as reviewer, could you pls have a quick look at the `jenkinsfile` and make sure we are ok? Thx!
